parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java

```diff
@@ -0,0 +1,351 @@
+    }
+
+    public byte[] getDataPageAAD() {
+      return this.dataPageAAD;
+    }
+
+    public byte[] getDictPageHeaderAAD() {
+      return this.dictPageHeaderAAD;
+    }
+
+    public byte[] getDictPageAAD() {
+      return this.dictPageAAD;
+    }
+  }
+
+  private final int pageBufferSize = ParquetProperties.DEFAULT_PAGE_SIZE * 2;
+  private byte[] pageBuffer;
+  private Configuration conf;
+
+  public ColumnEncryptor(Configuration conf) {
+    this.pageBuffer = new byte[pageBufferSize];
+    this.conf = conf;
+  }
+
+  /**
+   *
+   * @param inputFile Input file
+   * @param outputFile Output file
+   * @throws IOException
+   */
+  public void encryptColumns(String inputFile, String outputFile, List<String> paths, FileEncryptionProperties fileEncryptionProperties) throws IOException {
+    Path inPath = new Path(inputFile);
+    Path outPath = new Path(outputFile);
+
+    ParquetMetadata metaData = ParquetFileReader.readFooter(conf, inPath, NO_FILTER);
+    MessageType schema = metaData.getFileMetaData().getSchema();
+
+    ParquetFileWriter writer = new ParquetFileWriter(HadoopOutputFile.fromPath(outPath, conf), schema, ParquetFileWriter.Mode.OVERWRITE,
+      DEFAULT_BLOCK_SIZE, MAX_PADDING_SIZE_DEFAULT, DEFAULT_COLUMN_INDEX_TRUNCATE_LENGTH, DEFAULT_STATISTICS_TRUNCATE_LENGTH,
+      ParquetProperties.DEFAULT_PAGE_WRITE_CHECKSUM_ENABLED, fileEncryptionProperties);
+    writer.start();
+
+    try (TransParquetFileReader reader = new TransParquetFileReader(HadoopInputFile.fromPath(inPath, conf), HadoopReadOptions.builder(conf).build())) {
+      processBlocks(reader, writer, metaData, schema, paths);
+    } catch (Exception e) {
+      LOG.error("Exception happened while processing blocks: ", e);
+    } finally {
+      try {
+        writer.end(metaData.getFileMetaData().getKeyValueMetaData());
+      } catch (Exception e) {
+        LOG.error("Exception happened when ending the writer: ", e);
+      }
+    }
+  }
+
+  public void processBlocks(TransParquetFileReader reader, ParquetFileWriter writer, ParquetMetadata meta,
+                            MessageType schema, List<String> paths) throws IOException {
+    Set<ColumnPath> nullifyColumns = convertToColumnPaths(paths);
+    int blockId = 0;
+    PageReadStore store = reader.readNextRowGroup();
+
+    while (store != null) {
+      writer.startBlock(store.getRowCount());
+
+      List<ColumnChunkMetaData> columnsInOrder = meta.getBlocks().get(blockId).getColumns();
+      Map<ColumnPath, ColumnDescriptor> descriptorsMap = schema.getColumns().stream().collect(
+        Collectors.toMap(x -> ColumnPath.get(x.getPath()), x -> x));
+
+      for (int i = 0; i < columnsInOrder.size(); i += 1) {
+        ColumnChunkMetaData chunk = columnsInOrder.get(i);
+        ColumnDescriptor descriptor = descriptorsMap.get(chunk.getPath());
+        processChunk(descriptor, chunk, reader, writer, nullifyColumns, blockId, i, meta.getFileMetaData().getCreatedBy());
```

Comment:
Great idea! I am thinking if we introduce the universal tool, the file could be rewritten multiple times. How about "original.writer.version" and "rewriter. version", and the "rewriter. version" can be comma-separated and we can append to for every time rewriting. Let me know your thinking and I can implement it once we agree.

Reply:
Actually, I've suggested only one additional field and the other would be the already existing `created_by`. The question is which one should we add to `created_by` and which one to the key-value metadata in addition. Since we do not decode the data (so we cannot validate it) I think it is good to copy the `created_by` field from the original file and add the "rewriter version" to the key-value metadata. But it is a tricky question. Let's say we have a bug in one of the compression codecs so we generate invalid compressed pages with a certain version or parquet-mr. Now, we have a parquet file that we "trans-compress" the pages with a tool with this certain parquet-mr version and codec. Then, we use another tool to encrypt the file. I think this topic is more for parquet-format and the whole Parquet community. We shall introduce the discussed and accepted solution for any tool that "rewrites" a file while keeping the original data "undecoded".

Some questions (about a common rewriter) we need to answer before step forward:
* What would be the place of the rewriter versions? (New specific field or key-value metadata? Which key shall we use?)
* Shall we somehow also save what the rewriter has done? How?
* At what level shall we copy the original `created_by` field and what level shall we write the version of the rewriter to that field instead? (What different levels are possible?)
* From the introduction of this rewriter(s) field in case of any related writer version dependent fix we need to check this field as well and not only the `created_by` one.

Reply:
These are all great questions! I think we can start the conversation with some proposals and let the community review them. I just created the doc for the 'Unified Apache Rewriter' https://docs.google.com/document/d/1Ryt5uXnp-YwOrsnIDrGdTMoFfbOaTM6X39pBLHOa_50 and put all the questions into the open questions.

Reply:
I agree it is a bigger question. It is completely fine to handle separately from this PR. Meanwhile, we should ensure this won't get released before having the new field filled correctly. I would suggest having this tracked in a jira that is targeted for 1.13.0.

Reply:
https://issues.apache.org/jira/browse/PARQUET-2093 is created